File: apps/minecraft/spigot/src/main/java/org/leavesmc/leaves/protocol/jade/provider/entity/WaxedProvider.java
package org.leavesmc.leaves.protocol.jade.provider.entity;

import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.resources.Identifier;
import net.minecraft.util.Unit;
import net.minecraft.world.entity.animal.golem.CopperGolem;
import org.jadepaper.WidenedFields;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.leavesmc.leaves.protocol.jade.JadeProtocol;
import org.leavesmc.leaves.protocol.jade.accessor.EntityAccessor;
import org.leavesmc.leaves.protocol.jade.provider.StreamServerDataProvider;

public enum WaxedProvider implements StreamServerDataProvider<EntityAccessor, Unit> {
    INSTANCE;

    // Mirrors CopperGolem.IGNORE_WEATHERING_TICK: a waxed golem's oxidation is suspended,
    // so its next weathering tick is parked at this sentinel instead of a real game time.
    private static final long IGNORE_WEATHERING_TICK = -2L;
    private static final StreamCodec<RegistryFriendlyByteBuf, Unit> STREAM_CODEC = StreamCodec.unit(Unit.INSTANCE);
    private static final Identifier MC_WAXED = JadeProtocol.mc_id("waxed");

    @Override
    @Nullable
    public Unit streamData(@NotNull EntityAccessor accessor) {
        CopperGolem golem = (CopperGolem) accessor.getEntity();
        return WidenedFields.nextWeatheringTick(golem) == IGNORE_WEATHERING_TICK ? Unit.INSTANCE : null;
    }

    @Override
    public StreamCodec<RegistryFriendlyByteBuf, Unit> streamCodec() {
        return STREAM_CODEC;
    }

    @Override
    public Identifier getUid() {
        return MC_WAXED;
    }
}

File: apps/minecraft/spigot/src/main/kotlin/crabcraft/net/crabUtilities/jade/protocol/provider/entity/WaxedProvider.kt
package crabcraft.net.crabUtilities.jade.protocol.provider.entity

import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.resources.Identifier
import net.minecraft.util.Unit
import net.minecraft.world.entity.animal.golem.CopperGolem
import crabcraft.net.crabUtilities.jade.protocol.WidenedFields
import crabcraft.net.crabUtilities.jade.protocol.JadeProtocol
import crabcraft.net.crabUtilities.jade.protocol.accessor.EntityAccessor
import crabcraft.net.crabUtilities.jade.protocol.provider.StreamServerDataProvider

enum class WaxedProvider : StreamServerDataProvider<EntityAccessor, Unit> {
    INSTANCE;

    override fun streamData(accessor: EntityAccessor): Unit? {
        val golem = accessor.getEntity() as CopperGolem
        return if (WidenedFields.nextWeatheringTick(golem) == IGNORE_WEATHERING_TICK) Unit.INSTANCE else null
    }

    override fun streamCodec(): StreamCodec<RegistryFriendlyByteBuf, Unit> = STREAM_CODEC

    override fun getUid(): Identifier = MC_WAXED

    companion object {
        private val MC_WAXED = JadeProtocol.mc_id("waxed")
// Mirrors CopperGolem.IGNORE_WEATHERING_TICK: a waxed golem's oxidation is suspended,
        // so its next weathering tick is parked at this sentinel instead of a real game time.
        private const val IGNORE_WEATHERING_TICK = -2L
        private val STREAM_CODEC: StreamCodec<RegistryFriendlyByteBuf, Unit> = StreamCodec.unit(Unit.INSTANCE)
    }
}
